File: python/phone.py
#!/opt/bin/py
# -*- coding: utf-8 -*-

# & C:\Users\Max\AppData\Local\Programs\Python\Python39\python.exe -m pip install -U jinja2

from __future__ import print_function, absolute_import
from datetime import datetime, time, timedelta
#from jinja2 import Environment, PackageLoader, Template, select_autoescape

import math
import random

class PhoneTree:
  def __init__(self, type=0, iso=None):
    self.type = type
    self.iso = iso
    self.child = [None,None,None,None,None,None,None,None,None,None]

  def add(self, cc, iso):
    i = 0
    root = self
    for c in cc:
      i += 1
      index = ord(c) - ord('0')
      node = root.child[index]
      if node is None:
        root.child[index] = PhoneTree()
        node = root.child[index]
      if i < len(cc):
        root = node
      else:
        node.type = 1
        node.iso = iso

  def parse(self, phone):
    i = 0
    ln = 0
    ld = 0
    ac = 0
    iso = None

    root = self
    for c in phone:
      i += 1
      index = ord(c) - ord('0')
      node = root.child[index]
      if node is None:
        break
      if node.type == 1:
        if ln == 0:
          ln = i
        elif ld == 0:
          ld = i
        else:
          ac = i
        iso = node.iso
      root = node

    return (ln,ld,ac,iso)

File: lua/phone.lua

ippv = {child={}}

function ippv:new(o)
	o = o or {child={}}
	setmetatable(o, self)
	self.__index = self
	return o
end

function ippv:add(cc,iso)
	node = self
	for i = 1, string.len(cc) do
		index=string.byte(cc,i) - string.byte('0',1)
		if not node.child[index] then
			node.child[index] = {child={}}
		end
		node = node.child[index]
	end
	node.value = 1
	node.iso = iso
end

function ippv:parse(no)
	local node = self
	local cc = nil
	local ld = nil
	local ac = nil
	local iso = nil
	for i = 1, string.len(no) do
		index=string.byte(no,i) - string.byte('0',1)
		node = node.child[index]
		if not node then
			break
		end
		if node.value == nil or not node.value == 1 then
			goto continue
		end
		if not cc then
			cc = i
		elseif not ld then
			ld = i
		else
			ac = i
		end
		iso = node.iso
		::continue::
	end
	return cc,ld,ac,iso
end
